fingerprint2.js

```diff
@@ -21,1431 +21,1408 @@
-      if (this.options.excludeAudio) {
-        return done(keys)
-      }
+  }
 
-      if (this.options.excludeAudioIOS11 && navigator.userAgent.match(/OS 11.+Version\/11.+Safari/)) {
+  var isEnumerateDevicesSupported = function () {
+    return (navigator.mediaDevices && navigator.mediaDevices.enumerateDevices)
+  }
+// Inspired by and based on https://github.com/cozylife/audio-fingerprint
+  var audioKey = function (done, options) {
+    if (options.excludeAudioIOS11 && navigator.userAgent.match(/OS 11.+Version\/11.+Safari/)) {
         // See comment for excludeUserAgent and https://stackoverflow.com/questions/46363048/onaudioprocess-not-called-on-ios11#46534088
-        return done(keys)
-      }
+      return done(EXCLUDED)
+    }
 
-      var AudioContext = window.OfflineAudioContext || window.webkitOfflineAudioContext
+    var AudioContext = window.OfflineAudioContext || window.webkitOfflineAudioContext
 
-      if (AudioContext == null) {
-        keys.addPreprocessedComponent({key: 'audio_fp', value: null})
-        return done(keys)
-      }
+    if (AudioContext == null) {
+      return done(NOT_YET)
+    }
 
-      var context = new AudioContext(1, 44100, 44100)
+    var context = new AudioContext(1, 44100, 44100)
 
-      var oscillator = context.createOscillator()
-      oscillator.type = 'triangle'
-      oscillator.frequency.setValueAtTime(10000, context.currentTime)
+    var oscillator = context.createOscillator()
+    oscillator.type = 'triangle'
+    oscillator.frequency.setValueAtTime(10000, context.currentTime)
 
-      var compressor = context.createDynamicsCompressor()
-      this.each([
+    var compressor = context.createDynamicsCompressor()
+    each([
         ['threshold', -50],
         ['knee', 40],
         ['ratio', 12],
         ['reduction', -20],
         ['attack', 0],
         ['release', 0.25]
-      ], function (item) {
-        if (compressor[item[0]] !== undefined && typeof compressor[item[0]].setValueAtTime === 'function') {
-          compressor[item[0]].setValueAtTime(item[1], context.currentTime)
-        }
-      })
+    ], function (item) {
+      if (compressor[item[0]] !== undefined && typeof compressor[item[0]].setValueAtTime === 'function') {
+        compressor[item[0]].setValueAtTime(item[1], context.currentTime)
+      }
+    })
+
+    oscillator.connect(compressor)
+    compressor.connect(context.destination)
+    oscillator.start(0)
+    context.startRendering()
 
-      var audioTimeoutId = setTimeout(function () {
-        console.warn('Audio fingerprint timed out. Please report bug at https://github.com/Valve/fingerprintjs2 with your user agent: "' + navigator.userAgent + '".')
-        return done(keys)
-      }, 1000)
+    var audioTimeoutId = setTimeout(function () {
+      console.warn('Audio fingerprint timed out. Please report bug at https://github.com/Valve/fingerprintjs2 with your user agent: "' + navigator.userAgent + '".')
+      context.oncomplete = function () {}
```

Comment:
yes we can context = null, it will not make much difference, the real context will be eventually garbage collected when empty oncomplete function is called

Reply:
How about `.close()` or `.suspend()`?

Reply:
I don't know if such methods exists

Reply:
should be another issue to be honest